go-controller/pkg/ovn/loadbalancer/lb_cache.go

```diff
@@ -0,0 +1,376 @@
+			res.Protocol = str
+		} else if _, ok := row[2].([]interface{}); ok { //empty set
+			// empty protocol, assume tcp
+			res.Protocol = "tcp"
+		} else {
+			return nil, fmt.Errorf("failed to parse find load_balancer response: protocol: expected string, got %#v", row[2])
+		}
+
+		res.ExternalIDs, err = extractMap(row[3])
+		if err != nil {
+			return nil, fmt.Errorf("failed to parse find load_balancer response: external_ids: %w", err)
+		}
+
+		vips, err := extractMap(row[4])
+		if err != nil {
+			return nil, fmt.Errorf("failed to parse find load_balancer response: vips: %w", err)
+		}
+		for vip := range vips {
+			res.VIPs.Insert(vip)
+		}
+		out = append(out, res)
+	}
+
+	return out, nil
+}
+
+// findTableLBs returns a list of router name -> lb uuids
+func findTableLBs(tablename string) (map[string][]string, error) {
+	// this is CSV and not JSON because ovn-nbctl is inconsistent when a set has a single value :-/
+	rows, err := runNBCtlCSV([]string{"--data=bare", "--columns=name,load_balancer", "find", tablename})
+	if err != nil {
+		return nil, fmt.Errorf("failed to find existing LBs for %s: %w", tablename, err)
+	}
+
+	out := make(map[string][]string, len(rows))
+	for _, row := range rows {
+		if len(row) != 2 {
+			return nil, fmt.Errorf("invalid row returned when listing LBs for %s: %#v", tablename, row)
+		}
+		if row[0] == "" {
+			continue
+		}
+
+		lbs := strings.Split(row[1], " ")
+		if row[1] == "" {
+			lbs = []string{}
+		}
+		out[row[0]] = lbs
+	}
+
+	return out, nil
+}
+
+func TestOnlySetCache(cache *LBCache) {
+	globalCache = cache
+}
+
+// extractMap converts an untyped ovn json map in to a real map
+// it looks like this:
+// [ "map", [ ["k1", "v1"], ["k2", "v2"] ]]
+func extractMap(in interface{}) (map[string]string, error) {
+	out := map[string]string{}
+
+	// ["map", [ pairs]]
+	if cell, ok := in.([]interface{}); ok {
+		if len(cell) != 2 {
+			return nil, fmt.Errorf("expected outer pair, got %#v", cell)
+		} else {
+			// rows: [ [k,v], [k, v], ...]
+			if rows, ok := cell[1].([]interface{}); ok {
+				for _, row := range rows {
+					if pair, ok := row.([]interface{}); ok {
+						if len(pair) != 2 {
+							return nil, fmt.Errorf("expected k-v pair, got %#v", pair)
+						} else {
+							k, ok := pair[0].(string)
+							if !ok {
+								return nil, fmt.Errorf("key not a string: %#v", pair[0])
+							}
+							v, ok := pair[1].(string)
+							if !ok {
+								return nil, fmt.Errorf("value not a string: %#v", pair[1])
+							}
+
+							out[k] = v
+						}
+					} else {
+						return nil, fmt.Errorf("expected k-v slice, got %#v", row)
+					}
+				}
+			} else {
+				return nil, fmt.Errorf("expected map list, got %#v", cell[1])
+			}
+		}
+	} else {
+		return nil, fmt.Errorf("expected outer slice, got %#v", in)
+	}
+	return out, nil
+}
+
+// runNBCtlCSV runs an nbctl command that results in CSV output, parses the rows returned,
+// and returns the records
+func runNBCtlCSV(args []string) ([][]string, error) {
```

Comment:
should be in ovs.go with all the other commands

Reply:
Should it? It will go away with the move to libovsdb.

Reply:
i think so, it seems pretty generic to using nbctl

Reply:
done.